File: android/app/src/main/java/com/albunyaan/tube/ui/bootstrap/ProfileBootstrapViewModel.kt
package com.albunyaan.tube.ui.bootstrap

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.albunyaan.tube.auth.AccountRepository
import com.albunyaan.tube.auth.AgeIneligibleError
import com.albunyaan.tube.util.PhoneFormat
import com.google.firebase.auth.FirebaseAuth
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await

import java.time.LocalDate
import javax.inject.Inject

enum class BootstrapError {
    INVALID_NAME,
    INVALID_DOB,
    INVALID_PHONE_COUNTRY,
    INVALID_PHONE,
    INVALID_PASSWORD,
    PASSWORD_MISMATCH,
    PASSWORD_SET_FAILED,
    SAVE_FAILED,
}

sealed interface BootstrapNav {
    data object Idle : BootstrapNav
    data object NavigateToMain : BootstrapNav
    data object NavigateToAgeIneligible : BootstrapNav
}

@HiltViewModel
class ProfileBootstrapViewModel @Inject constructor(
    private val accountRepository: AccountRepository,
    private val firebaseAuth: FirebaseAuth,
    @ApplicationContext private val appContext: Context,
) : ViewModel() {

    data class UiState(
        val displayName: String = "",
        val dateOfBirth: LocalDate? = null,
        val phoneCountry: String? = null,    // ISO-3166-1 alpha-2 e.g. "NL"
        val phoneNumber: String = "",         // national-portion as typed
        val password: String = "",
        val passwordConfirm: String = "",
        /**
         * True when the signed-in Firebase user has only a Google provider
         * attached (no password). Setting a password during profile
         * completion attaches the password provider so the same email can
         * later sign in to the admin dashboard via email/password too —
         * the Android side of bidirectional auth (the web side handles
         * password→Google via account linking on LoginView).
         */
        val passwordRequired: Boolean = false,
        /**
         * True once accountRepository.completeProfile() has returned
         * success in this session. If the subsequent updatePassword call
         * fails, the user can retry without re-sending the profile data
         * (which the backend may reject as duplicate).
         */
        val profileSaved: Boolean = false,
        val isLoading: Boolean = false,
        val error: BootstrapError? = null,
    )

    private val _ui = MutableStateFlow(UiState())
    val ui: StateFlow<UiState> = _ui.asStateFlow()

    private val _nav = MutableStateFlow<BootstrapNav>(BootstrapNav.Idle)
    val nav: StateFlow<BootstrapNav> = _nav.asStateFlow()

    /**
     * Returns the first validation error in field-order, or null when the
     * state is valid. Shared source of truth between [isFormValid] (button
     * enable state) and [submit] (error-message dispatch) so the two cannot
     * drift out of sync.
     *
     * Requires [appContext] for libphonenumber, so lives on the ViewModel
     * rather than inside [UiState].
     */
    fun firstValidationError(s: UiState = _ui.value): BootstrapError? {
        val name = s.displayName.trim()
        if (name.isBlank() || name.length > 40) return BootstrapError.INVALID_NAME
        if (s.dateOfBirth == null)              return BootstrapError.INVALID_DOB
        if (s.phoneCountry.isNullOrBlank())     return BootstrapError.INVALID_PHONE_COUNTRY
        PhoneFormat.formatE164(appContext, s.phoneCountry, s.phoneNumber)
            ?: return BootstrapError.INVALID_PHONE
        if (s.passwordRequired) {
            if (s.password.length < MIN_PASSWORD_LENGTH) return BootstrapError.INVALID_PASSWORD
            if (s.password != s.passwordConfirm)         return BootstrapError.PASSWORD_MISMATCH
        }
        return null
    }

    /** Drives submit button enable state. True iff the form passes [firstValidationError]. */
    val isFormValid: Boolean get() = firstValidationError() == null

    fun seedDisplayName(initial: String) {
        if (_ui.value.displayName.isEmpty()) _ui.update { it.copy(displayName = initial) }
    }

    /**
     * Called by the Fragment after inspecting firebaseAuth.currentUser's
     * providerData. Drives the visibility of the password fields in the
     * UI — Google-only accounts get the prompt, email/password accounts
     * already have a password and shouldn't be asked again.
     */
    fun setPasswordRequirement(required: Boolean) {
        _ui.update { it.copy(passwordRequired = required) }
    }

    fun onDisplayNameChanged(v: String) {
        _ui.update { it.copy(displayName = v, error = null) }
    }

    fun onDobChanged(d: LocalDate) {
        _ui.update { it.copy(dateOfBirth = d, error = null) }
    }

    fun onPhoneCountryChanged(region: String) {
        _ui.update { it.copy(phoneCountry = region, error = null) }
    }

    fun onPhoneNumberChanged(v: String) {
        _ui.update { it.copy(phoneNumber = v, error = null) }
    }

    fun onPasswordChanged(v: String) {
        _ui.update { it.copy(password = v, error = null) }
    }

    fun onPasswordConfirmChanged(v: String) {
        _ui.update { it.copy(passwordConfirm = v, error = null) }
    }

    fun setLoading(loading: Boolean) {
        _ui.update { it.copy(isLoading = loading) }
    }

    fun surfaceError(e: BootstrapError) {
        _ui.update { it.copy(isLoading = false, error = e) }
    }

    fun consumeNav() { _nav.value = BootstrapNav.Idle }

    fun submit() {
        val s = _ui.value
        if (s.isLoading) return  // de-dupe rapid double-taps
        val validationError = firstValidationError(s)
        if (validationError != null) {
            _ui.update { it.copy(error = validationError) }
            return
        }
        val name = s.displayName.trim()
        val dob = s.dateOfBirth!!  // firstValidationError() guarantees non-null
        val phoneE164 = PhoneFormat.formatE164(appContext, s.phoneCountry!!, s.phoneNumber)!!
        _ui.update { it.copy(isLoading = true, error = null) }
        viewModelScope.launch {
            // Skip completeProfile if a prior submit already saved it and
            // we're only retrying the password step (which can fail
            // independently — see PASSWORD_SET_FAILED branch below).
            if (!s.profileSaved) {
                val profileResult = accountRepository.completeProfile(name, dob, phoneE164)
                profileResult.fold(
                    onSuccess = { _ui.update { it.copy(profileSaved = true) } },
                    onFailure = { e ->
                        _ui.update { it.copy(isLoading = false) }
                        if (e is AgeIneligibleError) {
                            _nav.value = BootstrapNav.NavigateToAgeIneligible
                        } else {
                            _ui.update { it.copy(error = BootstrapError.SAVE_FAILED) }
                        }
                        return@launch
                    }
                )
            }

            if (_ui.value.passwordRequired) {
                val user = firebaseAuth.currentUser
                if (user == null) {
                    // Session expired between profile save and password
                    // set. Profile is already committed backend-side; the
                    // user has to sign in again to attach a password.
                    _ui.update { it.copy(isLoading = false, error = BootstrapError.PASSWORD_SET_FAILED) }
                    return@launch
                }
                try {
                    user.updatePassword(_ui.value.password).await()
                } catch (e: Exception) {
                    // Profile is saved, password attach failed. Stay on
                    // screen so the user can retry password without
                    // re-submitting the profile (profileSaved guards
                    // against duplicate completeProfile calls).
                    _ui.update { it.copy(isLoading = false, error = BootstrapError.PASSWORD_SET_FAILED) }
                    return@launch
                }
            }

            _ui.update { it.copy(isLoading = false) }
            _nav.value = BootstrapNav.NavigateToMain
        }
    }

    companion object {
        const val MIN_PASSWORD_LENGTH = 8
    }
}

File: ios/FitrahTube/Features/Bootstrap/ProfileBootstrapViewModel.swift
import Foundation
import Observation

/// Drives `ProfileBootstrapScreen` (`ProfileBootstrapViewModel.kt`). Firebase stays behind
/// `AuthClient`, the backend behind `AccountClient`, and both the clock and the calendar are
/// injected so the age gate and the wire date are pure functions of what a test hands over.
@MainActor @Observable final class ProfileBootstrapViewModel {

    /// Where the screen goes once the form is settled. `.main` is advisory — `RootView` recomputes
    /// `SplashRouter.outcome` off the refreshed session and renders the shell itself.
    nonisolated enum Nav: Sendable, Equatable { case idle, main }

    nonisolated struct UiState: Equatable {
        var displayName = ""
        var dateOfBirth: Date?
        /// The NATIONAL portion as typed. The screen renders a fixed leading "+", so E.164 is
        /// assembled here (`e164`) rather than being something the user can get wrong.
        var phoneNumber = ""
        var password = ""
        var passwordConfirm = ""
        /// True when the signed-in account has no password provider: attaching one during bootstrap
        /// is what lets the same email later reach the admin dashboard from a browser
        /// (`ProfileBootstrapViewModel.kt:54-61`).
        var passwordRequired = false
        /// Set once `POST /profile` has returned 200. The backend 409s a second POST, so a retry
        /// after a failed password attach must re-run ONLY the password step.
        var profileSaved = false
        var isLoading = false
        var error: BootstrapError?
    }

    private let account: AccountClient
    private let auth: any AuthClient
    private let session: AccountSession
    private let calendar: Calendar
    private let today: @Sendable () -> Date

    private(set) var state = UiState()
    private(set) var nav: Nav = .idle

    init(account: AccountClient, auth: any AuthClient, session: AccountSession,
         calendar: Calendar = .current, today: @escaping @Sendable () -> Date = { Date() }) {
        self.account = account
        self.auth = auth
        self.session = session
        self.calendar = calendar
        self.today = today
    }

    // MARK: - Bindings

    // Every setter clears the standing error: a message about the previous attempt has nothing to
    // say about the text now on screen (`onDisplayNameChanged` and friends).
    /// Capped at 40 (`android:maxLength="40"`), so a paste is capped exactly as typing is and the
    /// server's `@Size(max = 40)` can never be the thing that reports it.
    var displayName: String {
        get { state.displayName }
        set {
            state.displayName = String(newValue.prefix(BootstrapValidator.maxNameLength))
            state.error = nil
        }
    }

    var dateOfBirth: Date? {
        get { state.dateOfBirth }
        set { state.dateOfBirth = newValue; state.error = nil }
    }

    var phoneNumber: String {
        get { state.phoneNumber }
        // `BootstrapValidator.normalizedDigits` — the same rule `EditPhoneSheet` applies, spelled
        // once beside the pattern it feeds (Stage 1 / B3a).
        set {
            state.phoneNumber = BootstrapValidator.normalizedDigits(newValue)
            state.error = nil
        }
    }

    var password: String {
        get { state.password }
        set { state.password = newValue; state.error = nil }
    }

    var passwordConfirm: String {
        get { state.passwordConfirm }
        set { state.passwordConfirm = newValue; state.error = nil }
    }

    /// E.164 as the server sees it.
    var e164: String { BootstrapValidator.e164(state.phoneNumber) }

    // MARK: - Validation (ONE validator, both consumers)

    func firstError() -> BootstrapError? {
        BootstrapValidator.firstError(name: state.displayName, dob: state.dateOfBirth, phone: e164,
                                      password: state.password, passwordConfirm: state.passwordConfirm,
                                      passwordRequired: state.passwordRequired,
                                      today: today(), calendar: calendar)
    }

    /// Drives the submit button's enabled state.
    var isFormValid: Bool { firstError() == nil }

    // MARK: - Lifecycle

    /// The screen's `.task`. `currentUser()` is async on iOS, so the password requirement cannot be
    /// derived at construction the way Android's fragment does it.
    func load() async {
        guard let user = await auth.currentUser() else { return }
        state.passwordRequired = !user.hasPasswordProvider
    }

    /// Two-phase commit. Phase one is `POST /profile`, latched by `profileSaved`; phase two attaches
    /// the password. A failure in phase two leaves the user on this screen with the latch set, so the
    /// retry sends only the password (`ProfileBootstrapViewModel.kt:161-217`).
    func submit() async {
        guard !state.isLoading else { return }   // de-dupe rapid double-taps
        if let error = firstError() {
            state.error = error
            return
        }
        guard let dob = state.dateOfBirth else { return }   // firstError() guarantees non-nil
        state.isLoading = true
        state.error = nil

        if !state.profileSaved {
            do {
                _ = try await account.completeProfile(
                    displayName: state.displayName.trimmingCharacters(in: .whitespacesAndNewlines),
                    dateOfBirth: Self.wireDate(dob, calendar: calendar),
                    phoneNumber: e164)
                state.profileSaved = true
            } catch {
                // Stage 3 / M4: both of these were decoded, thrown, tested — and then collapsed
                // into "couldn't save your profile", which for the 409 is a dead end BY
                // CONSTRUCTION: the server says the form is already done, and the only thing that
                // could move the user on is the `/me` re-read this arm never issued, so the screen
                // repeated the same refusal forever. A 403 `EMAIL_NOT_VERIFIED` is the same shape —
                // the router lands on verification once the session is re-read.
                switch error {
                case .profileAlreadyCompleted, .emailNotVerified:
                    await session.refresh()
                    state.isLoading = false
                    nav = .main
                    return
                case .ageIneligible:
                    state.isLoading = false
                    // R7-P1 #3: the teardown runs WITH the verdict, not when the user acknowledges
                    // it. The server revokes the refresh tokens and DISABLES the Firebase account
                    // before answering (`AccountProfileService.java:130,140`), so a session kept
                    // alive past this point is one every later request 401s on -- and the refused
                    // forced mint maps `.userDisabled` to `.blocked`, which replaced the age
                    // message with "your account has been blocked" and re-routed off `.signedOut`
                    // so the Firebase delete never ran at all. `RootView` presents the terminal
                    // screen on `session.isAgeIneligible`, over whatever the outcome now resolves
                    // to; the terminal screen IS the message, so no inline error either.
                    await session.terminateAgeIneligible()
                    return
                default:
                    state.isLoading = false
                    state.error = .saveFailed
                    return
                }
            }
        }

        if state.passwordRequired {
            // The session can expire between the two phases. The profile is already committed, so
            // the only way back is a fresh sign-in — say what happened, never why.
            guard await auth.currentUser() != nil else {
                state.isLoading = false
                state.error = .passwordSetFailed
                return
            }
            do {
                try await auth.updatePassword(state.password)
            } catch {
                state.isLoading = false
                state.error = .passwordSetFailed
                return
            }
        }

        // `RootView` routes off `AccountSession.state.me?.status` and NOTHING else re-reads `/me`, so
        // without this the account stays `pending_profile` and the screen it was just released from
        // renders again.
        await session.refresh()
        state.isLoading = false
        nav = .main
    }

    /// ISO `yyyy-MM-dd`, in ASCII digits, for the day the picker SHOWED.
    ///
    /// `String(format:)` takes no locale and so never localizes its digits — a `DateFormatter` on
    /// `Locale.current` would emit Arabic-Indic digits for an `ar` user and the backend would 400 on
    /// every submit. The TIME ZONE is the picker's (`Calendar.current` in the app), NOT UTC: a user
    /// at UTC+13 picking 2000-01-01 holds an instant that is still 1999-12-31 in UTC, and sending
    /// that would make every such account a day younger than it is. The NUMBERING is always
    /// Gregorian (`BootstrapValidator.gregorian`) — the picker may render Hijri, the wire may not.
    nonisolated static func wireDate(_ date: Date, calendar: Calendar) -> String {
        let parts = BootstrapValidator.gregorian(calendar).dateComponents([.year, .month, .day], from: date)
        return String(format: "%04d-%02d-%02d", parts.year ?? 0, parts.month ?? 0, parts.day ?? 0)
    }
}

extension BootstrapError {
    /// The inline message under the form. `passwordSetFailed`/`saveFailed` say WHAT, never why.
    var messageKey: String {
        switch self {
        case .invalidName: "bootstrap_error_invalid_name"
        case .invalidDOB: "bootstrap_error_invalid_dob"
        case .underAge: "bootstrap_error_under_age"
        case .invalidPhone: "bootstrap_error_invalid_phone"
        case .invalidPassword: "bootstrap_error_invalid_password"
        case .passwordMismatch: "bootstrap_error_password_mismatch"
        case .passwordSetFailed: "bootstrap_error_password_set_failed"
        case .saveFailed: "bootstrap_error_save_failed"
        }
    }
}
